segmentstore/server/src/main/java/io/pravega/segmentstore/server/attributes/SegmentAttributeIndex.java

```diff
@@ -634,11 +857,190 @@ private void read00(RevisionDataInput input, AttributeCollectionBuilder builder)
+                        // in order to pass it down for serialization.
+                        values.put(e.getKey(), new UUID(version, e.getValue()));
+                        changed |= existing == null || existing.getLeastSignificantBits() != e.getValue();
+                    }
+                }
+            }
+
+            // Serialize, but only if anything changed.
+            if (changed) {
+                byte[] newData = CacheEntryLayout.setValues(
+                        existingData,
+                        values.entrySet().stream().sorted(Comparator.comparing(Map.Entry::getKey)).iterator(),
+                        values.size());
+                SegmentAttributeIndex.this.cache.insert(getKey(), newData);
+                this.size = newData.length;
+                this.generation = currentGeneration;
+            }
+        }
+
+        /**
+         * Removes all data associated with this cache entry.
+         */
+        synchronized void clear() {
+            SegmentAttributeIndex.this.cache.remove(getKey());
+            this.size = 0;
+        }
+
+        /**
+         * Reads the values from the given serialization for the given Attribute Ids and populates them into the given Map.
+         *
+         * @param data         The Cache Entry serialization.
+         * @param attributeIds A Collection of Attribute Ids to fetch.
+         * @param result       A Map to store the result. Only those Attribute Ids that exist in data will be stored here.
+         * @return True if at least one value was fetched, false otherwise.
+         */
+        private boolean readValues(byte[] data, Collection<UUID> attributeIds, Map<UUID, Long> result) {
+            // The Cache Entry Serialization has the Attributes in sorted order (by Id), so the most efficient way of
+            // searching the values is to do a binary search within the byte array:
+            // - We make use of the CacheEntryLayout abstraction to not have to deserialize the whole array.
+            // - We process the incoming attributeIds in sorted order, which will aid our search. We remember the index
+            // of the last found Attribute Id, and only execute our search for subsequent Ids from that index on, since,
+            // if those Ids exist, they are guaranteed to be after the last found one.
+            final int count = CacheEntryLayout.getCount(data);
+
+            // Process the Attribute Ids in sorted order.
+            int nextPos = 0; // Position of the last found Attribute Id. This is a sequence position, and not the index in the array.
+            boolean anythingFound = false;
+            Iterator<UUID> iterator = attributeIds.stream().sorted().iterator();
+            while (iterator.hasNext() && nextPos < count) {
+                // Do a binary search starting at nextIndex until the end of the data.
```

Comment:
Please abstract this out.

Reply:
Into what? This is already isolated as it is, and is highly specialized to the problem at hand. I don't see any other use for this, and trying to generalize it too much would involve too much overhead and thus make it less efficient.

Reply:
I see your point. I extracted binarySearch out.